File: src/emby_range_cache_proxy/cache.py
from __future__ import annotations

import hashlib
import os
import re
import time
import uuid
from pathlib import Path

from .models import ByteRange, MediaSource, SourceMetadata

GIB = 1024**3
MIB = 1024**2
KEY_PATTERN = re.compile(r"^[0-9a-f]{64}$")


def adaptive_head_tail(size: int) -> tuple[int, int]:
    if size < 2 * GIB:
        return 16 * MIB, 4 * MIB
    if size < 8 * GIB:
        return 32 * MIB, 8 * MIB
    if size <= 30 * GIB:
        return 64 * MIB, 8 * MIB
    return 128 * MIB, 16 * MIB


def cache_key(source: MediaSource, metadata: SourceMetadata) -> str:
    material = "\n".join(
        [
            source.media_source_id,
            metadata.url,
            str(metadata.size),
            metadata.etag or "",
            metadata.last_modified or "",
        ]
    )
    return hashlib.sha256(material.encode("utf-8")).hexdigest()


class HeadTailCache:
    def __init__(self, root: str | Path, *, max_bytes: int) -> None:
        self.root = Path(root)
        self.max_bytes = max_bytes
        self.root.mkdir(parents=True, exist_ok=True)

    def block_path(self, key: str, block_name: str) -> Path:
        return self._cache_dir(key) / f"{self._validate_block_name(block_name)}.bin"

    def meta_path(self, key: str, block_name: str) -> Path:
        return self._cache_dir(key) / f"{self._validate_block_name(block_name)}.range"

    def store_block(self, key: str, block_name: str, byte_range: ByteRange, data: bytes) -> None:
        if len(data) != byte_range.length:
            raise ValueError("data length must match byte_range length")
        directory = self._cache_dir(key)
        directory.mkdir(parents=True, exist_ok=True)
        path = self.block_path(key, block_name)
        tmp = path.with_name(f"{path.name}.tmp")
        tmp.write_bytes(data)
        os.replace(tmp, path)
        meta = self.meta_path(key, block_name)
        meta_tmp = meta.with_name(f"{meta.name}.tmp")
        meta_tmp.write_text(f"{byte_range.start}-{byte_range.end}\n")
        os.replace(meta_tmp, meta)
        self._touch(path)

    def stage_block(self, key: str, block_name: str, byte_range: ByteRange) -> "CacheBlockWriter":
        return CacheBlockWriter(self, key, block_name, byte_range)

    def read_block(self, key: str, block_name: str, requested: ByteRange) -> bytes | None:
        path = self.block_path(key, block_name)
        meta = self.meta_path(key, block_name)
        if not path.exists() or not meta.exists():
            return None
        try:
            stored = self._read_range(meta)
        except (OSError, ValueError):
            self._remove_entry(path, meta)
            return None
        if requested.start < stored.start or requested.end > stored.end:
            return None
        with path.open("rb") as handle:
            handle.seek(requested.start - stored.start)
            data = handle.read(requested.length)
        if len(data) != requested.length:
            self._remove_entry(path, meta)
            return None
        self._touch(path)
        return data

    def evict_if_needed(self) -> None:
        files = [path for path in self.root.glob("*/*.bin") if path.is_file()]
        total = sum(path.stat().st_size for path in files)
        if total <= self.max_bytes:
            return
        files.sort(key=lambda path: path.stat().st_mtime_ns)
        for path in files:
            if total <= self.max_bytes:
                break
            size = path.stat().st_size
            path.unlink(missing_ok=True)
            path.with_suffix(".range").unlink(missing_ok=True)
            total -= size

    def _read_range(self, path: Path) -> ByteRange:
        start, end = path.read_text().strip().split("-", 1)
        return ByteRange(int(start), int(end))

    def _touch(self, path: Path) -> None:
        now = time.time_ns()
        os.utime(path, ns=(now, now))

    def _cache_dir(self, key: str) -> Path:
        if not KEY_PATTERN.fullmatch(key):
            raise ValueError("cache key must be 64 lowercase hex characters")
        return self.root / key

    def _validate_block_name(self, block_name: str) -> str:
        if block_name not in {"head", "tail"}:
            raise ValueError("block_name must be head or tail")
        return block_name

    def _remove_entry(self, path: Path, meta: Path) -> None:
        path.unlink(missing_ok=True)
        meta.unlink(missing_ok=True)


class CacheBlockWriter:
    def __init__(self, cache: HeadTailCache, key: str, block_name: str, byte_range: ByteRange) -> None:
        self.cache = cache
        self.key = key
        self.block_name = cache._validate_block_name(block_name)
        self.byte_range = byte_range
        self.directory = cache._cache_dir(key)
        self.directory.mkdir(parents=True, exist_ok=True)
        self.path = cache.block_path(key, block_name)
        self.meta = cache.meta_path(key, block_name)
        self.tmp = self.path.with_name(f"{self.path.name}.{uuid.uuid4().hex}.tmp")
        self._handle = self.tmp.open("wb")
        self.bytes_written = 0
        self._closed = False

    def write(self, data: bytes) -> None:
        if self._closed:
            raise ValueError("cache block writer is closed")
        self._handle.write(data)
        self.bytes_written += len(data)

    def commit(self) -> None:
        if self._closed:
            raise ValueError("cache block writer is closed")
        self._handle.close()
        self._closed = True
        if self.bytes_written != self.byte_range.length:
            self.tmp.unlink(missing_ok=True)
            raise ValueError("staged data length must match byte_range length")
        meta_tmp = self.meta.with_name(f"{self.meta.name}.{uuid.uuid4().hex}.tmp")
        meta_tmp.write_text(f"{self.byte_range.start}-{self.byte_range.end}\n")
        os.replace(self.tmp, self.path)
        os.replace(meta_tmp, self.meta)
        self.cache._touch(self.path)

    def abort(self) -> None:
        if not self._closed:
            self._handle.close()
            self._closed = True
        self.tmp.unlink(missing_ok=True)

    def __enter__(self) -> "CacheBlockWriter":
        return self

    def __exit__(self, exc_type, exc, tb) -> None:
        if exc_type is not None or not self._closed:
            self.abort()

File: go/internal/headtail/cache.go
package headtail

import (
	"fmt"
	"io"
	"os"
	"path/filepath"
	"regexp"
	"sort"
	"strconv"
	"strings"
	"time"

	"github.com/xmm2022/emby-range-cache-proxy/go/internal/model"
)

var keyRE = regexp.MustCompile(`^[0-9a-f]{64}$`)

type Cache struct {
	Root     string
	MaxBytes int64
}

func NewCache(root string, maxBytes int64) *Cache {
	_ = os.MkdirAll(root, 0o755)
	return &Cache{Root: root, MaxBytes: maxBytes}
}

func (c *Cache) StoreBlock(key, blockName string, byteRange model.ByteRange, data []byte) error {
	if int64(len(data)) != byteRange.Length() {
		return fmt.Errorf("data length must match byte range")
	}
	writer, err := c.StageBlock(key, blockName, byteRange)
	if err != nil {
		return err
	}
	if _, err := writer.Write(data); err != nil {
		writer.Abort()
		return err
	}
	return writer.Commit()
}

func (c *Cache) StageBlock(key, blockName string, byteRange model.ByteRange) (*BlockWriter, error) {
	if err := validateKey(key); err != nil {
		return nil, err
	}
	if err := validateBlockName(blockName); err != nil {
		return nil, err
	}
	if byteRange.Start < 0 || byteRange.End < byteRange.Start {
		return nil, fmt.Errorf("invalid byte range")
	}
	dir := filepath.Join(c.Root, key)
	if err := os.MkdirAll(dir, 0o755); err != nil {
		return nil, err
	}
	tmp, err := os.CreateTemp(dir, blockName+".*.tmp")
	if err != nil {
		return nil, err
	}
	return &BlockWriter{
		cache:     c,
		key:       key,
		blockName: blockName,
		byteRange: byteRange,
		tmpPath:   tmp.Name(),
		handle:    tmp,
	}, nil
}

func (c *Cache) IterBlock(key, blockName string, requested model.ByteRange, chunkBytes int64) (<-chan []byte, error) {
	if chunkBytes <= 0 {
		return nil, fmt.Errorf("chunkBytes must be positive")
	}
	if err := validateKey(key); err != nil {
		return nil, err
	}
	if err := validateBlockName(blockName); err != nil {
		return nil, err
	}
	path := c.blockPath(key, blockName)
	meta := c.metaPath(key, blockName)
	stored, err := readRange(meta)
	if err != nil {
		c.removeEntry(path, meta)
		return nil, nil
	}
	stat, err := os.Stat(path)
	if err != nil || stat.Size() != stored.Length() {
		c.removeEntry(path, meta)
		return nil, nil
	}
	if requested.Start < stored.Start || requested.End > stored.End {
		return nil, nil
	}
	handle, err := os.Open(path)
	if err != nil {
		c.removeEntry(path, meta)
		return nil, nil
	}
	if _, err := handle.Seek(requested.Start-stored.Start, io.SeekStart); err != nil {
		handle.Close()
		c.removeEntry(path, meta)
		return nil, nil
	}
	_ = touch(path)
	out := make(chan []byte)
	go func() {
		defer close(out)
		defer handle.Close()
		remaining := requested.Length()
		buf := make([]byte, int(chunkBytes))
		for remaining > 0 {
			n := int64(len(buf))
			if n > remaining {
				n = remaining
			}
			readBuf := buf[:n]
			if _, err := io.ReadFull(handle, readBuf); err != nil {
				c.removeEntry(path, meta)
				return
			}
			chunk := make([]byte, n)
			copy(chunk, readBuf)
			out <- chunk
			remaining -= n
		}
	}()
	return out, nil
}

func (c *Cache) HasBlockRange(key, blockName string, requested model.ByteRange) (bool, error) {
	if err := validateKey(key); err != nil {
		return false, err
	}
	if err := validateBlockName(blockName); err != nil {
		return false, err
	}
	path := c.blockPath(key, blockName)
	meta := c.metaPath(key, blockName)
	stored, err := readRange(meta)
	if err != nil {
		c.removeEntry(path, meta)
		return false, nil
	}
	stat, err := os.Stat(path)
	if err != nil || stat.Size() != stored.Length() {
		c.removeEntry(path, meta)
		return false, nil
	}
	return requested.Start >= stored.Start && requested.End <= stored.End, nil
}

func (c *Cache) EvictIfNeeded() error {
	if c.MaxBytes <= 0 {
		return nil
	}
	files, err := filepath.Glob(filepath.Join(c.Root, "*", "*.bin"))
	if err != nil {
		return err
	}
	type fileInfo struct {
		path  string
		size  int64
		mtime int64
	}
	infos := make([]fileInfo, 0, len(files))
	var total int64
	for _, path := range files {
		stat, err := os.Stat(path)
		if err != nil || stat.IsDir() {
			continue
		}
		total += stat.Size()
		infos = append(infos, fileInfo{path: path, size: stat.Size(), mtime: stat.ModTime().UnixNano()})
	}
	sort.Slice(infos, func(i, j int) bool { return infos[i].mtime < infos[j].mtime })
	for _, info := range infos {
		if total <= c.MaxBytes {
			break
		}
		_ = os.Remove(info.path)
		_ = os.Remove(strings.TrimSuffix(info.path, ".bin") + ".range")
		total -= info.size
	}
	return nil
}

func (c *Cache) blockPath(key, blockName string) string {
	return filepath.Join(c.Root, key, blockName+".bin")
}

func (c *Cache) metaPath(key, blockName string) string {
	return filepath.Join(c.Root, key, blockName+".range")
}

func (c *Cache) removeEntry(path, meta string) {
	_ = os.Remove(path)
	_ = os.Remove(meta)
}

type BlockWriter struct {
	cache        *Cache
	key          string
	blockName    string
	byteRange    model.ByteRange
	tmpPath      string
	handle       *os.File
	bytesWritten int64
	closed       bool
}

func (w *BlockWriter) Write(data []byte) (int, error) {
	if w.closed {
		return 0, fmt.Errorf("writer is closed")
	}
	n, err := w.handle.Write(data)
	w.bytesWritten += int64(n)
	return n, err
}

func (w *BlockWriter) Commit() error {
	if w.closed {
		return fmt.Errorf("writer is closed")
	}
	if err := w.handle.Close(); err != nil {
		w.closed = true
		_ = os.Remove(w.tmpPath)
		return err
	}
	w.closed = true
	if w.bytesWritten != w.byteRange.Length() {
		_ = os.Remove(w.tmpPath)
		return fmt.Errorf("staged data length must match byte range")
	}
	path := w.cache.blockPath(w.key, w.blockName)
	meta := w.cache.metaPath(w.key, w.blockName)
	metaTmp := meta + "." + filepath.Base(w.tmpPath) + ".tmp"
	if err := os.WriteFile(metaTmp, []byte(fmt.Sprintf("%d-%d\n", w.byteRange.Start, w.byteRange.End)), 0o644); err != nil {
		_ = os.Remove(w.tmpPath)
		return err
	}
	if err := os.Rename(w.tmpPath, path); err != nil {
		_ = os.Remove(w.tmpPath)
		_ = os.Remove(metaTmp)
		return err
	}
	if err := os.Rename(metaTmp, meta); err != nil {
		_ = os.Remove(path)
		_ = os.Remove(metaTmp)
		return err
	}
	return touch(path)
}

func (w *BlockWriter) Abort() {
	if !w.closed {
		_ = w.handle.Close()
		w.closed = true
	}
	_ = os.Remove(w.tmpPath)
}

func BlockForRequest(byteRange model.ByteRange, size, headSize, tailSize int64) (string, model.ByteRange) {
	headEnd := headSize
	if headEnd > size {
		headEnd = size
	}
	head := model.ByteRange{Start: 0, End: headEnd - 1}
	tailStart := size - tailSize
	if tailStart < 0 {
		tailStart = 0
	}
	tail := model.ByteRange{Start: tailStart, End: size - 1}
	if contains(head, byteRange) {
		return "head", head
	}
	if contains(tail, byteRange) {
		return "tail", tail
	}
	return "", model.ByteRange{}
}

func contains(container, requested model.ByteRange) bool {
	return requested.Start >= container.Start && requested.End <= container.End
}

func validateKey(key string) error {
	if !keyRE.MatchString(key) {
		return fmt.Errorf("cache key must be 64 lowercase hex characters")
	}
	return nil
}

func validateBlockName(blockName string) error {
	if blockName != "head" && blockName != "tail" {
		return fmt.Errorf("block name must be head or tail")
	}
	return nil
}

func readRange(path string) (model.ByteRange, error) {
	data, err := os.ReadFile(path)
	if err != nil {
		return model.ByteRange{}, err
	}
	parts := strings.Split(strings.TrimSpace(string(data)), "-")
	if len(parts) != 2 {
		return model.ByteRange{}, fmt.Errorf("invalid range sidecar")
	}
	start, err := strconv.ParseInt(parts[0], 10, 64)
	if err != nil {
		return model.ByteRange{}, err
	}
	end, err := strconv.ParseInt(parts[1], 10, 64)
	if err != nil || end < start {
		return model.ByteRange{}, fmt.Errorf("invalid range sidecar")
	}
	return model.ByteRange{Start: start, End: end}, nil
}

func touch(path string) error {
	now := time.Now()
	return os.Chtimes(path, now, now)
}
